Multiple parents

AZWN opened the pull request:
This PR adds the option to specify multiple parents for a `ResourceService`.

AZWN committed:
Implement multiple parents for ResourceService

AZWN committed:
Add additional methods to create child resource services

MeAmAnUsername commented on src/main/java/mb/resource/ResourceService.java:
```diff
@@ -305,4 +320,28 @@ default ResourceService createChild(Iterable<ResourceRegistry> registries) {
+     * @param otherParents  Other resource services that serve as parent for the child.
+     * @return Child resource service.
+     */
+    default ResourceService createChild(Iterable<ResourceRegistry> registries, Iterable<ResourceService> otherParents) {
```
`otherParents` is not used

AZWN replied:
:heavy_check_mark: Thnx!

AZWN committed:
Add forgotten parameter

MeAmAnUsername commented on src/main/java/mb/resource/DefaultResourceService.java:
```diff
@@ -224,15 +232,19 @@ public HierarchicalResource appendOrReplaceWithHierarchical(HierarchicalResource
+    public DefaultResourceService createChild(ResourceRegistry defaultRegistry, Iterable<ResourceRegistry> registries, Iterable<ResourceService> otherParents) {
+        ArrayList<ResourceService> parents = new ArrayList<>();
+        parents.add(this);
+        otherParents.forEach(parents::add);
```
Given how the previous implementation worked, I don't think you are supposed to add the parents of this `ResourceService` to the child. I think the parents of this `ResourceService` should be grandparents of the child, like a linked list of parents.
I don't know for sure though, I don't know the codebase very well.

AZWN replied:
I don't know the codebase very well either, allthough I choose this implementation because:
- It reflects the mental model of the design the best
- The intended parallel change in Pie follows the same pattern.

However, we might hear Gabriels opinion about this.

Gohla replied:
Both styles should work because ResourceService is immutable. However, maybe the field should be called `ancestors` in this case to make the intent clear?

AZWN replied:
I've pushed the change.

Gohla replied:
I don't see any changes

AZWN replied:
Annoying. I pushed the changed (verified [here](https://github.com/AZWN/resource/tree/multiple-parents())), but it does not appear on this PR. I'll try to figure out.

AZWN replied:
Repushed the commit, and now it shows up.

AZWN committed:
Remove unnecessary method call/parameter

AZWN committed:
Rename parents to ancestors

In documentation, retained 'parent' for single parent, and used 'ancestor' when multiple parents are used.

Gohla reviewed (approved):
Looks good, thanks.
